File: frontend/src/shared/config/shiftTypes.ts
/**
 * Справочник типов смен.
 *
 * Источник истины для UI — фронт мапит code → (label, color, letter) локально.
 * Бэкенд хранит только `shift_type_code VARCHAR(20)`, см. backend/app/core/shift_types.py.
 *
 * Синхронизация при изменении: правки в обоих файлах в одном коммите.
 */

export interface ShiftTypeMeta {
  code: string
  name: string
  startTime: string | null
  endTime: string | null
  plannedHours: number
  isWorking: boolean
  isNight: boolean
  color: string
  letter: string | null
  sortOrder: number
}

export const SHIFT_TYPE_CATALOG: ShiftTypeMeta[] = [
  { code: "day",       name: "День (08:00-16:30)",      startTime: "08:00", endTime: "16:30", plannedHours: 8,  isWorking: true,  isNight: false, color: "#10b981", letter: null,  sortOrder: 10 },
  { code: "day_long",  name: "День 12ч (08:00-20:00)",  startTime: "08:00", endTime: "20:00", plannedHours: 12, isWorking: true,  isNight: false, color: "#22c55e", letter: null,  sortOrder: 20 },
  { code: "night",     name: "Ночь 12ч (20:00-08:00)",  startTime: "20:00", endTime: "08:00", plannedHours: 12, isWorking: true,  isNight: true,  color: "#1e3a8a", letter: null,  sortOrder: 30 },
  { code: "evening",   name: "Вечер (14:00-22:00)",     startTime: "14:00", endTime: "22:00", plannedHours: 8,  isWorking: true,  isNight: false, color: "#f59e0b", letter: null,  sortOrder: 60 },
  { code: "off",       name: "Выходной",                startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#9ca3af", letter: "В",   sortOrder: 100 },
  { code: "vacation",  name: "Отпуск",                  startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#3b82f6", letter: "О",   sortOrder: 110 },
  { code: "sick",      name: "Больничный",              startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#ef4444", letter: "Б",   sortOrder: 120 },
  { code: "A",         name: "За свой счет",            startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#a855f7", letter: "А",   sortOrder: 125 },
  { code: "D",         name: "Донорские",               startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#14b8a6", letter: "Д",   sortOrder: 135 },
  { code: "absence",   name: "Прогул / Неявка",         startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#dc2626", letter: "П",   sortOrder: 140 },
  { code: "VK",        name: "Военкомат",               startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#059669", letter: "ВК",  sortOrder: 145 },
  { code: "VS",        name: "Военные сборы",           startTime: null,    endTime: null,    plannedHours: 0,  isWorking: false, isNight: false, color: "#047857", letter: "ВС",  sortOrder: 150 },
]

const BY_CODE: Record<string, ShiftTypeMeta> = Object.fromEntries(
  SHIFT_TYPE_CATALOG.map((s) => [s.code, s])
)

export function getShiftTypeMeta(code: string | null | undefined): ShiftTypeMeta | null {
  if (!code) return null
  return BY_CODE[code] ?? null
}

File: backend/app/core/shift_types.py
"""Справочник типов смен.

Хранится в коде, а не в БД: набор типов фиксирован (не редактируется пользователем),
и UI-метаданные (название, времена, иконка, цвет) — это front-end concern.

Изменение списка = релиз (код + фронт), без миграций и сидинга.

Фронт мапит code → (label, icon, color) по этому же каталогу
(см. frontend/src/shared/config/shiftTypes.ts).
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import time
from typing import Dict, List, Optional


@dataclass(frozen=True)
class ShiftType:
    code: str
    name: str
    start_time: Optional[time]
    end_time: Optional[time]
    planned_hours: float
    is_working: bool
    is_night: bool
    sort_order: int


def _t(hh: int, mm: int) -> time:
    return time(hh, mm)


# Контрактные смены завода. times подобраны под фактический график.
SHIFT_TYPE_CATALOG: List[ShiftType] = [
    ShiftType("day", "День (08:00-16:30)", _t(8, 0), _t(16, 30), 8.0, True, False, 10),
    ShiftType("day_long", "День 12ч (08:00-20:00)", _t(8, 0), _t(20, 0), 12.0, True, False, 20),
    ShiftType("night", "Ночь 12ч (20:00-08:00)", _t(20, 0), _t(8, 0), 12.0, True, True, 30),
    ShiftType("evening", "Вечер (14:00-22:00)", _t(14, 0), _t(22, 0), 8.0, True, False, 60),
    ShiftType("off", "Выходной", None, None, 0.0, False, False, 100),
    ShiftType("vacation", "Отпуск", None, None, 0.0, False, False, 110),
    ShiftType("sick", "Больничный", None, None, 0.0, False, False, 120),
    ShiftType("A", "За свой счет", None, None, 0.0, False, False, 125),
    ShiftType("D", "Донорские", None, None, 0.0, False, False, 135),
    ShiftType("absence", "Прогул / Неявка", None, None, 0.0, False, False, 140),
    ShiftType("VK", "Военкомат", None, None, 0.0, False, False, 145),
    ShiftType("VS", "Военные сборы", None, None, 0.0, False, False, 150),
]


_VALID_CODES: Dict[str, ShiftType] = {st.code: st for st in SHIFT_TYPE_CATALOG}


def get_shift_type(code: Optional[str]) -> Optional[ShiftType]:
    if not code:
        return None
    return _VALID_CODES.get(code)


def validate_shift_type_code(code: Optional[str]) -> Optional[str]:
    """Возвращает код если он валиден, иначе None. None тоже допустим (пустая смена)."""
    if code is None or code == "":
        return None
    if code not in _VALID_CODES:
        raise ValueError(f"Неизвестный тип смены: {code!r}")
    return code
